WebGLRenderingContextBase preventBufferClearForInspector is redundant

kkinnunen-apple opened the pull request:
#### 27dac32ecc9c989f5d4cd702baba2db6fc23fdde
<pre>
WebGLRenderingContextBase preventBufferClearForInspector is redundant
<a href="https://bugs.webkit.org/show_bug.cgi?id=263885">https://bugs.webkit.org/show_bug.cgi?id=263885</a>
<a href="https://rdar.apple.com/117684773">rdar://117684773</a>

Reviewed by Dan Glastonbury.

WebGLRenderingContextBase::setPreventBufferClearForInspector
was intended to make sure that inspector reading the WebGL rendering
would not force the pending drawing buffer clear. The read happens
with toDataURL, which generally used to read drawing buffer, not
compositing buffer.

This was a workaround for case where the read would happen after
prepareForDisplay. However, preventing the clear is incorrect, as
there is nothing correct to be read from the uncleared back buffer.
That buffer may contain contents from two frames back or it might
be a brand new buffer.

Instead, make it explicit that the transfer to the rendering results
happens from display buffer if there is no draws in drawing
buffer. Otherwise use the drawing buffer.

Remove the redundant m_isDisplayingWithPaint, instead prepare
the display buffer during paint, if it is not already prepared. Normal
layer paint happens after document-wide preparation during rendering
update. Snapshots paint or print may happen without preparation, so
the HTMLCanvasElement::paint will prepare the display buffer.

Removes the canvas reading taint check logic from Web Inspector capture.
The capture should always work, as it is trusted. Each individual
draw commands will be captured differently, and there the result
data reflects what the target canvas receives, taint logic included.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::paint):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::prepareForDisplayWithPaint): Deleted.
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::prepareForDisplayWithPaint): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::initializeContextState):
(WebCore::WebGLRenderingContextBase::markContextChanged):
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::paintRenderingResultsToCanvas):
(WebCore::WebGLRenderingContextBase::prepareForDisplayWithPaint): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::preventBufferClearForInspector const): Deleted.
(WebCore::WebGLRenderingContextBase::setPreventBufferClearForInspector): Deleted.
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::getContentAsDataURL):

Canonical link: <a href="https://commits.webkit.org/270046@main">https://commits.webkit.org/270046@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cfec6b5e7f7d159a22898851d16812320a05f3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24425 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2867 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26556 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22469 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/24696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4511 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/214 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24671 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/2040 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21099 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27142 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/1776 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22024 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28228 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/22264 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22345 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26021 "Found 2 new API test failures: /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/dom-cache, /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/origin-and-total-storage-ratio (failure)") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/1728 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/55 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/3013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5841 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2158 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/2101 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebCore/html/OffscreenCanvas.cpp:
```diff
@@ -539,8 +539,9 @@ void OffscreenCanvas::commitToPlaceholderCanvas()
     if (m_context && (m_context->isWebGL() || m_context->isAccelerated())) {
-        m_context->prepareForDisplayWithPaint();
-        m_context->paintRenderingResultsToCanvas();
+        if (m_context->compositingResultsNeedUpdating())
+            m_context->prepareForDisplay();
```
Should this logic be part of `m_context->drawBufferToCanvas`?

kkinnunen-apple replied:
No, the buffer swap logic is explicit, where as drawBufferToCanvas can just select which one to draw

djg commented on Source/WebCore/html/canvas/CanvasRenderingContext.h:
```diff
@@ -79,10 +79,17 @@ class CanvasRenderingContext : public ScriptWrappable, public CanMakeWeakPtr<Can
-    virtual void prepareForDisplayWithPaint() { }
-    virtual void paintRenderingResultsToCanvas() { }
+    // Canvas 2DContext drawing buffer is the same as display buffer.
+    // WebGL, WebGPU draws to drawing buffer and composites to display buffer.
```
Nits: Is the display buffer composited to, or is it the source that is composited? Is something like:
```
// WebGL, WebGPU draws to the drawing buffer and the display buffer is sent to the compositor.
```
???

kkinnunen-apple replied:
Thanks!

djg commented on Source/WebCore/html/canvas/CanvasRenderingContext.h:
```diff
@@ -91,6 +98,7 @@ class CanvasRenderingContext : public ScriptWrappable, public CanMakeWeakPtr<Can
 
     virtual bool compositingResultsNeedUpdating() const { return false; }
     virtual bool needsPreparationForDisplay() const { return false; }
+    // Swaps the current drawing buffer to display buffer.
     virtual void prepareForDisplay() { }
```
This comment makes me think the function is incorrectly named.

kkinnunen-apple replied:
You're right. It might be prudent to change it to some sort of swapbuffers in future commits

djg commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -568,6 +568,10 @@ static String ensureNotNull(const String& text)
 
+static GraphicsContextGL::SurfaceBuffer toGCGLSurfaceBuffer(CanvasRenderingContext::SurfaceBuffer buffer)
+{
+    return buffer == CanvasRenderingContext::SurfaceBuffer::DrawingBuffer ? GraphicsContextGL::SurfaceBuffer::DrawingBuffer : GraphicsContextGL::SurfaceBuffer::DisplayBuffer;
```
Is the `CanvasRenderingContext::SurfaceBuffer`/`GraphicsContext::SurfaceBuffer` duplication really needed?

kkinnunen-apple replied:
It's a bit inconvenient, but I'd like to preserve it (GraphicsContextGL is lower level than CanvasRenderingContext, and CanvasRenderingContext cannot depend on GraphicsContextGL). Having a freestanding SurfaceBuffer seems a bit odd and specific and disconnected. If there's a good name, maybe freestanding enum would be the best.

djg reviewed (approved):
:shipit:

kkinnunen-apple committed:
WebGLRenderingContextBase preventBufferClearForInspector is redundant
https://bugs.webkit.org/show_bug.cgi?id=263885
rdar://117684773

Reviewed by Dan Glastonbury.

WebGLRenderingContextBase::setPreventBufferClearForInspector
was intended to make sure that inspector reading the WebGL rendering
would not force the pending drawing buffer clear. The read happens
with toDataURL, which generally used to read drawing buffer, not
compositing buffer.

This was a workaround for case where the read would happen after
prepareForDisplay. However, preventing the clear is incorrect, as
there is nothing correct to be read from the uncleared back buffer.
That buffer may contain contents from two frames back or it might
be a brand new buffer.

Instead, make it explicit that the transfer to the rendering results
happens from display buffer if there is no draws in drawing
buffer. Otherwise use the drawing buffer.

Remove the redundant m_isDisplayingWithPaint, instead prepare
the display buffer during paint, if it is not already prepared. Normal
layer paint happens after document-wide preparation during rendering
update. Snapshots paint or print may happen without preparation, so
the HTMLCanvasElement::paint will prepare the display buffer.

Removes the canvas reading taint check logic from Web Inspector capture.
The capture should always work, as it is trusted. Each individual
draw commands will be captured differently, and there the result
data reflects what the target canvas receives, taint logic included.

* Source/WebCore/html/HTMLCanvasElement.cpp:
(WebCore::HTMLCanvasElement::paint):
* Source/WebCore/html/OffscreenCanvas.cpp:
(WebCore::OffscreenCanvas::commitToPlaceholderCanvas):
* Source/WebCore/html/canvas/CanvasRenderingContext.h:
(WebCore::CanvasRenderingContext::prepareForDisplayWithPaint): Deleted.
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.h:
* Source/WebCore/html/canvas/GPUCanvasContextCocoa.mm:
(WebCore::GPUCanvasContextCocoa::prepareForDisplayWithPaint): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::initializeContextState):
(WebCore::WebGLRenderingContextBase::markContextChanged):
(WebCore::WebGLRenderingContextBase::clearIfComposited):
(WebCore::WebGLRenderingContextBase::paintRenderingResultsToCanvas):
(WebCore::WebGLRenderingContextBase::prepareForDisplayWithPaint): Deleted.
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
(WebCore::WebGLRenderingContextBase::preventBufferClearForInspector const): Deleted.
(WebCore::WebGLRenderingContextBase::setPreventBufferClearForInspector): Deleted.
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::getContentAsDataURL):

Canonical link: https://commits.webkit.org/270046@main
